serverless.js

```diff
@@ -0,0 +1,112 @@
+const { Component } = require('@serverless/core')
+const { equals, is, isEmpty, isNil, mergeDeepRight, not } = require('ramda')
+
+const {
+  constructTemplateS3Key,
+  createOrUpdateStack,
+  deleteStack,
+  fetchOutputs,
+  getClients,
+  getPreviousStack,
+  loadTemplate,
+  uploadTemplate,
+  updateTerminationProtection
+} = require('./utils')
+
+const defaults = {
+  enableTerminationProtection: false,
+  parameters: {},
+  region: 'us-east-1',
+  rollbackConfiguration: {}
+}
+
+class AwsCloudFormation extends Component {
+  async default(inputs = {}) {
+    this.context.status('Deploying')
+    const config = mergeDeepRight(defaults, inputs)
+    config.externalBucket = not(isNil(config.bucket))
+    config.bucket = this.state.bucket || config.bucket
+    config.timestamp = Date.now()
+
+    if (is(String, config.template)) {
+      this.context.debug('Load template file.')
+      config.template = await loadTemplate(config.template)
+    }
+
+    if (isNil(config.template) || isNil(config.stackName)) {
+      throw new Error('Invalid inputs; template and stackName are required.')
+    }
+
+    config.templateS3Key = constructTemplateS3Key(config)
+
+    if (isNil(config.bucket)) {
+      const awsS3 = await this.load('@serverless/aws-s3')
+      this.context.debug('Creating S3 bucket for the template.')
+      const { name } = await awsS3()
+      config.bucket = name
+    }
+
+    const { cloudformation, s3 } = getClients(this.context.credentials.aws, config.region)
+
+    if (not(isNil(this.state.stackName)) && not(equals(this.state.stackName, config.stackName))) {
+      this.context.debug(
+        `Delete stack ${this.state.stackName} prior to creation of ${config.stackName}.`
+      )
+      await deleteStack(cloudformation, this.state)
+    }
```

Comment:
since this step will probably take some time, and that the new stack does not depend on the old one, I'm thinking maybe it would be a lot faster to delete the old one, and create the new one in parallel using `Promise.all` or something. What you think? 🤔 

Also, I always delete the old resource in case of a rename at the end **after** the new one has been created, just in case something goes wrong with creation. I think that could apply here too. Does that make sense?

Reply:
I was thinking that the resource names might collide and the deployment fails because of that. But yes, it is better to remove old resources after the new ones are deployed.

Reply:
✅fixed